pkg/cmd/run.go

```diff
@@ -241,82 +205,242 @@ doppler run --mount secrets.json -- cat secrets.json`,
-		} else {
-			// remove any reserved keys from secrets
-			reservedKeys := []string{"PATH", "PS1", "HOME"}
-			for _, reservedKey := range reservedKeys {
-				if _, found := dopplerSecrets[reservedKey]; found == true {
-					utils.LogDebug(fmt.Sprintf("Ignoring reserved secret %s", reservedKey))
-					delete(dopplerSecrets, reservedKey)
+			// this lock ensures the old process, if any, has exited before we start a new process
+			processMutex.Lock()
+
+			// we could have received a new update event while we were waiting for the previous process to terminate.
+			// if so, don't bother starting the process as it'll just be immediately restarted again after fetching the latest secrets
+			if lastUpdateEvent.After(secretsFetchedAt) {
+				utils.LogDebug("Not starting new process; more recent update event has been received")
+				processMutex.Unlock()
+				return
+			}
+
+			terminatedByWatch = false
+
+			var env []string
+			env, cleanupMount = controllers.PrepareSecrets(secrets, originalEnv, preserveEnv, existingEnvKeys, mountOptions)
+
+			global.WaitGroup.Add(1)
+
+			if isRestart {
+				utils.Log("Restarting process")
+			}
+
+			// start the process
+			c, err = controllers.Run(cmd, args, env, forwardSignals)
+			if err != nil {
+				defer global.WaitGroup.Done()
+				if cleanupMount != nil {
+					cleanupMount()
 				}
+				utils.HandleError(err)
 			}
 
-			if preserveEnv != "false" {
-				secretsToPreserve := strings.Split(preserveEnv, ",")
+			go func() {
+				defer processMutex.Unlock()
+				defer global.WaitGroup.Done()
 
-				// use doppler secrets
-				for name, value := range dopplerSecrets {
-					secrets[name] = value
+				exitCode, err := utils.WaitCommand(c)
+
+				if cleanupMount != nil {
+					cleanupMount()
+
+					// add arbitrary delay before re-mounting secrets to avoid a race-related "broken pipe" when writing to the named pipe
+					time.Sleep(100 * time.Millisecond)
 				}
-				// then use existing env vars
-				for name, value := range existingEnvKeys {
-					if preserveEnv != "true" && !utils.Contains(secretsToPreserve, name) {
-						continue
-					}
 
-					if _, found := secrets[name]; found == true {
-						utils.LogDebug(fmt.Sprintf("Ignoring Doppler secret %s", name))
+				// ignore errors if we were responsible for killing the process
+				if !terminatedByWatch {
+					if err != nil {
+						if strings.HasPrefix(err.Error(), "exec") || strings.HasPrefix(err.Error(), "fork/exec") {
+							utils.LogError(err)
+						}
+						utils.LogDebugError(err)
 					}
-					secrets[name] = value
-				}
-			} else {
-				// use existing env vars
-				for name, value := range existingEnvKeys {
-					secrets[name] = value
+
+					os.Exit(exitCode)
 				}
-				// then use doppler secrets
-				for name, value := range dopplerSecrets {
-					secrets[name] = value
+			}()
+		}
+
+		watchHandler := func(data []byte) {
+			// Expected format: "event: message\ndata: {JSON}\n\n"
```

Comment:
recommendation/option: What do you think about pulling the event parsing into a helper func? I think it's making this function a little too heavy

Reply:
Good call